File: scripts/run_feii_bin_stability.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Run the canonical Fe II log-cosine bin-stability ladder.

Bins: 120, 160, 200
Default null-n: 500 (quick reproducibility). Use --null-n 5000 for full.

Outputs go to:
    outputs/fe_ion2_120/
    outputs/fe_ion2_160/
    outputs/fe_ion2_200/

Master summary is written to:
    tables/nist_master_results.csv

If a canonical output folder already contains nist_summary.json, that run is
reused unless --force is supplied. No old run is deleted.
"""

from __future__ import annotations

import argparse
import csv
import json
import subprocess
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional

REPO_ROOT = Path(__file__).resolve().parent.parent
SCANNER = REPO_ROOT / "scripts" / "nist_wct_log_spectral_scan_FIXED.py"
FE_CSV = REPO_ROOT / "data" / "Fe_lines.csv"
OUT_ROOT = REPO_ROOT / "outputs"
TABLES_DIR = REPO_ROOT / "tables"
MASTER_CSV = TABLES_DIR / "nist_master_results.csv"

BINS_LADDER = [120, 160, 200]

MASTER_COLUMNS = [
    "species",
    "ion",
    "bins",
    "n_unique_lines",
    "k_best",
    "n_obs",
    "deltaD",
    "scan_null_p",
    "tail_count_ge",
    "null_n",
    "ell_min",
    "ell_max",
    "delta_ell",
    "out_dir",
    "status",
]


def load_summary(out_dir: Path) -> Optional[Dict[str, Any]]:
    p = out_dir / "nist_summary.json"
    if not p.exists():
        return None
    with open(p, "r", encoding="utf-8") as f:
        return json.load(f)


def summary_to_row(species: str, ion: int, bins: int, out_dir: Path, status: str) -> Dict[str, Any]:
    row = {c: "" for c in MASTER_COLUMNS}
    row.update({
        "species": species,
        "ion": ion,
        "bins": bins,
        "out_dir": str(out_dir.relative_to(REPO_ROOT)) if out_dir.is_absolute() else str(out_dir),
        "status": status,
    })
    s = load_summary(out_dir)
    if s:
        best = s.get("best", {}) or {}
        br = s.get("branch_report", {}) or {}
        row["n_unique_lines"] = s.get("n_unique_lines", "")
        row["k_best"] = best.get("k_best", "")
        row["n_obs"] = br.get("n_obs", "")
        row["deltaD"] = best.get("deltaD", "")
        row["scan_null_p"] = s.get("scan_null_p", "")
        row["tail_count_ge"] = s.get("tail_count_ge", "")
        row["null_n"] = s.get("null_n", "")
        row["ell_min"] = s.get("ell_min", "")
        row["ell_max"] = s.get("ell_max", "")
        row["delta_ell"] = s.get("delta_ell", "")
    return row


def run_scanner(bins: int, null_n: int, out_dir: Path, min_lines: int, seed: int, n_k: int) -> int:
    cmd = [
        sys.executable, str(SCANNER),
        "--csv", str(FE_CSV),
        "--ion", "2",
        "--bins", str(bins),
        "--null-n", str(null_n),
        "--min-lines", str(min_lines),
        "--out-dir", str(out_dir),
        "--n-k", str(n_k),
        "--seed", str(seed),
    ]
    print("[run]", " ".join(cmd))
    completed = subprocess.run(cmd, check=False)
    return completed.returncode


def write_master(rows: List[Dict[str, Any]]) -> None:
    TABLES_DIR.mkdir(parents=True, exist_ok=True)
    with open(MASTER_CSV, "w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=MASTER_COLUMNS)
        w.writeheader()
        for r in rows:
            w.writerow({c: r.get(c, "") for c in MASTER_COLUMNS})
    print(f"[save] {MASTER_CSV.relative_to(REPO_ROOT)}")


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--null-n", type=int, default=500)
    ap.add_argument("--min-lines", type=int, default=100)
    ap.add_argument("--n-k", type=int, default=2500)
    ap.add_argument("--seed", type=int, default=20260517)
    ap.add_argument("--force", action="store_true",
                    help="Re-run scanner even if canonical output already exists. "
                         "Existing files are overwritten by the scanner inside that "
                         "canonical folder; non-canonical legacy folders are never touched.")
    args = ap.parse_args()

    if not SCANNER.exists():
        print(f"[error] scanner not found: {SCANNER}", file=sys.stderr)
        return 2
    if not FE_CSV.exists():
        print(f"[error] Fe CSV not found: {FE_CSV}", file=sys.stderr)
        return 2

    rows: List[Dict[str, Any]] = []
    OUT_ROOT.mkdir(parents=True, exist_ok=True)

    for bins in BINS_LADDER:
        out_dir = OUT_ROOT / f"fe_ion2_{bins}"
        out_dir.mkdir(parents=True, exist_ok=True)
        existing = load_summary(out_dir)
        if existing and not args.force:
            print(f"[reuse] {out_dir.relative_to(REPO_ROOT)} (nist_summary.json present)")
            status = "reused"
        else:
            rc = run_scanner(bins, args.null_n, out_dir, args.min_lines, args.seed, args.n_k)
            status = "ok" if rc == 0 else f"failed_rc{rc}"
        rows.append(summary_to_row("Fe", 2, bins, out_dir, status))

    write_master(rows)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: R/run_feii_bin_stability.R
#!/usr/bin/env Rscript
# run_feii_bin_stability.R
# ---------------------------------------------------------------------------
# Run the Fe II log-cosine bin-stability ladder (bins 120, 160, 200) with the
# R scanner and aggregate the results into an R-generated master table:
#
#     tables_r/nist_master_results_r.csv
#
# R outputs go to outputs_r/fe_ion2_{120,160,200}/ so that the canonical Python
# results under outputs/ are never touched. If a folder already contains
# nist_summary.json that run is reused unless --force is supplied.
#
# Canonical Python reference values (regression targets, NOT hard-coded here):
#   120 bins: k_best=31.3265306122449  n_obs~10.717134  deltaD~355.7443
#   160 bins: k_best=31.3265306122449  n_obs~10.739649  deltaD~315.7277
#   200 bins: k_best=31.3265306122449  n_obs~10.753158  deltaD~259.1788
# ---------------------------------------------------------------------------

suppressWarnings(suppressMessages(library(jsonlite)))

.invoked_file <- sub("^--file=", "", commandArgs(trailingOnly = FALSE)[grep("^--file=", commandArgs(trailingOnly = FALSE))])
SCRIPT_DIR <- if (length(.invoked_file) > 0L) dirname(normalizePath(.invoked_file)) else "R"
REPO_ROOT <- normalizePath(file.path(SCRIPT_DIR, ".."))

SCANNER <- file.path(SCRIPT_DIR, "nist_wct_log_spectral_scan.R")
FE_CSV <- file.path(REPO_ROOT, "data", "Fe_lines.csv")
OUT_ROOT <- file.path(REPO_ROOT, "outputs_r")
TABLES_DIR <- file.path(REPO_ROOT, "tables_r")
MASTER_CSV <- file.path(TABLES_DIR, "nist_master_results_r.csv")

BINS_LADDER <- c(120L, 160L, 200L)
MASTER_COLUMNS <- c(
  "species", "ion", "bins", "n_unique_lines", "k_best", "n_obs", "deltaD",
  "scan_null_p", "tail_count_ge", "null_n", "ell_min", "ell_max", "delta_ell",
  "out_dir", "status"
)

parse_args <- function(argv) {
  vals <- list(`null-n` = 500L, `min-lines` = 100L, `n-k` = 2500L,
               seed = 20260517L, force = FALSE)
  i <- 1L
  while (i <= length(argv)) {
    key <- sub("^--", "", argv[i])
    if (key == "force") {
      vals$force <- TRUE
      i <- i + 1L
    } else if (key %in% names(vals)) {
      vals[[key]] <- if (key == "seed" || key %in% c("null-n", "min-lines", "n-k")) {
        as.integer(argv[i + 1L])
      } else {
        argv[i + 1L]
      }
      i <- i + 2L
    } else {
      stop(sprintf("Unknown option: %s", argv[i]))
    }
  }
  vals
}

load_summary <- function(out_dir) {
  p <- file.path(out_dir, "nist_summary.json")
  if (!file.exists(p)) return(NULL)
  jsonlite::fromJSON(p)
}

summary_to_row <- function(species, ion, bins, out_dir, status) {
  row <- setNames(as.list(rep("", length(MASTER_COLUMNS))), MASTER_COLUMNS)
  row$species <- species
  row$ion <- ion
  row$bins <- bins
  row$out_dir <- sub(paste0("^", REPO_ROOT, "/?"), "", out_dir)
  row$status <- status
  s <- load_summary(out_dir)
  if (!is.null(s)) {
    row$n_unique_lines <- s$n_unique_lines
    row$k_best <- s$best$k_best
    row$n_obs <- s$branch_report$n_obs
    row$deltaD <- s$best$deltaD
    row$scan_null_p <- if (is.null(s$scan_null_p)) "" else s$scan_null_p
    row$tail_count_ge <- if (is.null(s$tail_count_ge)) "" else s$tail_count_ge
    row$null_n <- s$null_n
    row$ell_min <- s$ell_min
    row$ell_max <- s$ell_max
    row$delta_ell <- s$delta_ell
  }
  row
}

run_scanner <- function(bins, null_n, out_dir, min_lines, seed, n_k) {
  cmd_args <- c(
    SCANNER,
    "--csv", FE_CSV,
    "--ion", "2",
    "--bins", as.character(bins),
    "--null-n", as.character(null_n),
    "--min-lines", as.character(min_lines),
    "--out-dir", out_dir,
    "--n-k", as.character(n_k),
    "--seed", as.character(seed)
  )
  cat("[run] Rscript", paste(cmd_args, collapse = " "), "\n")
  status <- system2("Rscript", cmd_args, stdout = "", stderr = "")
  status
}

write_master <- function(rows) {
  dir.create(TABLES_DIR, showWarnings = FALSE, recursive = TRUE)
  df <- do.call(rbind, lapply(rows, function(r) {
    as.data.frame(r[MASTER_COLUMNS], stringsAsFactors = FALSE)
  }))
  # write full numeric precision
  utils::write.csv(df, MASTER_CSV, row.names = FALSE)
  cat(sprintf("[save] %s\n", sub(paste0("^", REPO_ROOT, "/?"), "", MASTER_CSV)))
}

main <- function(argv = commandArgs(trailingOnly = TRUE)) {
  args <- parse_args(argv)
  if (!file.exists(SCANNER)) stop(sprintf("scanner not found: %s", SCANNER))
  if (!file.exists(FE_CSV)) stop(sprintf("Fe CSV not found: %s", FE_CSV))

  dir.create(OUT_ROOT, showWarnings = FALSE, recursive = TRUE)
  rows <- list()
  for (bins in BINS_LADDER) {
    out_dir <- file.path(OUT_ROOT, sprintf("fe_ion2_%d", bins))
    dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
    existing <- load_summary(out_dir)
    if (!is.null(existing) && !args$force) {
      cat(sprintf("[reuse] %s (nist_summary.json present)\n",
                  sub(paste0("^", REPO_ROOT, "/?"), "", out_dir)))
      status <- "reused"
    } else {
      rc <- run_scanner(bins, args[["null-n"]], out_dir,
                        args[["min-lines"]], args$seed, args[["n-k"]])
      status <- if (rc == 0L) "ok" else sprintf("failed_rc%d", rc)
    }
    rows[[length(rows) + 1L]] <- summary_to_row("Fe", 2L, bins, out_dir, status)
  }
  write_master(rows)
  invisible(0L)
}

if (length(.invoked_file) > 0L && grepl("run_feii_bin_stability\\.R$", .invoked_file)) {
  main()
}
